pkg/store/cache/caching_bucket.go

```diff
@@ -0,0 +1,389 @@
+		return 0, err
+	}
+
+	var buf [8]byte
+	binary.BigEndian.PutUint64(buf[:], size)
+	cb.cache.Store(ctx, map[string][]byte{key: buf[:]}, ttl)
+
+	return size, nil
+}
+
+func (cb *CachingBucket) getRangeChunkFile(ctx context.Context, name string, offset, length int64) (io.ReadCloser, error) {
+	size, err := cb.cachedObjectSize(ctx, name, cb.config.ChunkObjectSizeTTL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get size of chunk file: %s", name)
+	}
+
+	// If length goes over object size, adjust length. We use it later to limit number of read bytes.
+	if uint64(offset+length) > size {
+		length = int64(size - uint64(offset))
+	}
+
+	startOffset := (offset / cb.config.ChunkBlockSize) * cb.config.ChunkBlockSize
+	endOffset := ((offset + length) / cb.config.ChunkBlockSize) * cb.config.ChunkBlockSize
+	if (offset+length)%cb.config.ChunkBlockSize > 0 {
+		endOffset += cb.config.ChunkBlockSize
+	}
+
+	blocks := (endOffset - startOffset) / cb.config.ChunkBlockSize
+
+	offsetKeys := make(map[int64]string, blocks)
+	keys := make([]string, 0, blocks)
+	lastBlockLength := 0
+
+	for o := startOffset; o < endOffset; o += cb.config.ChunkBlockSize {
+		end := o + cb.config.ChunkBlockSize
+		if end > int64(size) {
+			end = int64(size)
+		}
+
+		lastBlockLength = int(end - o)
+
+		k := cachingKeyObjectBlock(name, o, end)
+		keys = append(keys, k)
+		offsetKeys[o] = k
+	}
+
+	// Try to get all blocks from the cache.
+	hits, _ := cb.cache.Fetch(ctx, keys)
+
+	for _, b := range hits {
+		cb.cachedChunkBytes.Add(float64(len(b)))
+	}
+
+	if len(hits) < len(keys) {
+		if hits == nil {
+			hits = map[string][]byte{}
+		}
+
+		err := cb.fetchMissingChunkBlocks(ctx, name, startOffset, endOffset, offsetKeys, hits, lastBlockLength)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return newBlocksReader(cb.config.ChunkBlockSize, offsetKeys, hits, offset, length), nil
+}
+
+type rng struct {
+	start, end int64
+}
+
+// Fetches missing blocks and stores them into "hits" map.
+func (cb *CachingBucket) fetchMissingChunkBlocks(ctx context.Context, name string, startOffset, endOffset int64, cacheKeys map[int64]string, hits map[string][]byte, lastBlockLength int) error {
+	// Ordered list of missing sub-ranges.
+	var missing []rng
+
+	for off := startOffset; off < endOffset; off += cb.config.ChunkBlockSize {
+		if hits[cacheKeys[off]] == nil {
+			missing = append(missing, rng{start: off, end: off + cb.config.ChunkBlockSize})
+		}
+	}
+
+	missing = mergeRanges(missing, 0) // Merge adjacent ranges.
+	// Keep merging until we have only max number of ranges (= requests).
+	for limit := cb.config.ChunkBlockSize; cb.config.MaxChunksGetRangeRequests > 0 && len(missing) > cb.config.MaxChunksGetRangeRequests; limit = limit * 2 {
+		missing = mergeRanges(missing, limit)
+	}
+
+	var hitsMutex sync.Mutex
+
+	// Run parallel queries for each missing range. Fetched data is stored into 'hits' map, protected by hitsMutex.
+	eg, nctx := errgroup.WithContext(ctx)
```

Comment:
To keep consistency with the rest of the code I would suggest to rename `eg` to `g` and `nctx` to `gctx`